src/shims/sync.rs

```diff
@@ -454,135 +456,325 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+        mutex_set_id(this, mutex_op, ScalarMaybeUndef::Undef)?;
 
         Ok(0)
     }
 
     fn pthread_rwlock_rdlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
-        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
+        let id = rwlock_get_or_create_id(this, rwlock_op)?;
+        let active_thread = this.get_active_thread()?;
 
-        if writers != 0 {
-            // The lock is locked by a writer.
-            assert_eq!(writers, 1);
-            let reader_blockset = rwlock_get_or_create_reader_blockset(this, rwlock_op)?;
-            this.block_active_thread(reader_blockset)?;
+        if this.rwlock_is_write_locked(id) {
+            this.rwlock_enqueue_reader(id, active_thread);
+            this.block_thread(active_thread)?;
             Ok(0)
         } else {
-            match readers.checked_add(1) {
-                Some(new_readers) => {
-                    rwlock_set_readers(this, rwlock_op, Scalar::from_u32(new_readers))?;
-                    Ok(0)
-                }
-                None => this.eval_libc_i32("EAGAIN"),
-            }
+            this.rwlock_reader_add(id, active_thread);
+            Ok(0)
         }
     }
 
     fn pthread_rwlock_tryrdlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
-        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
-        if writers != 0 {
+        let id = rwlock_get_or_create_id(this, rwlock_op)?;
+        let active_thread = this.get_active_thread()?;
+
+        if this.rwlock_is_write_locked(id) {
             this.eval_libc_i32("EBUSY")
         } else {
-            match readers.checked_add(1) {
-                Some(new_readers) => {
-                    rwlock_set_readers(this, rwlock_op, Scalar::from_u32(new_readers))?;
-                    Ok(0)
-                }
-                None => this.eval_libc_i32("EAGAIN"),
-            }
+            this.rwlock_reader_add(id, active_thread);
+            Ok(0)
         }
     }
 
     fn pthread_rwlock_wrlock(&mut self, rwlock_op: OpTy<'tcx, Tag>) -> InterpResult<'tcx, i32> {
         let this = self.eval_context_mut();
 
-        let readers = rwlock_get_readers(this, rwlock_op)?.to_u32()?;
-        let writers = rwlock_get_writers(this, rwlock_op)?.to_u32()?;
-        let writer_blockset = rwlock_get_or_create_writer_blockset(this, rwlock_op)?;
-        if readers != 0 || writers != 0 {
-            this.block_active_thread(writer_blockset)?;
+        let id = rwlock_get_or_create_id(this, rwlock_op)?;
+        let active_thread = this.get_active_thread()?;
+
+        if this.rwlock_is_locked(id) {
+            this.block_thread(active_thread)?;
```

Comment:
This will deadlock if the current thread already holds the lock, right? Please add a link to https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_rwlock_wrlock.html to show that that is the right behavior. (That [was not always clear](https://github.com/rust-lang/rust/issues/53127).)

Reply:
Yes. It is tested in `libc_pthread_rwlock_write_read_deadlock` and `libc_pthread_rwlock_write_write_deadlock`.